idaes/apps/grid_integration/multiperiod/price_taker_model.py

```diff
@@ -0,0 +1,1008 @@
+            raise ImportError(
+                f"Clustering feature requires optional import 'scikit-learn'."
+            )
+        kmeans = KMeans(n_clusters=n_clusters).fit(daily_data.transpose())
+        centroids = kmeans.cluster_centers_
+        labels = kmeans.labels_
+
+        # Set any centroid values that are < 1e-4 to 0 to avoid noise
+        centroids = centroids * (abs(centroids) >= 1e-4)
+
+        # Compute weight for each cluster by counting its occurrences in the dataset
+        unique_labels, weights_counter = np.unique(labels, return_counts=True)
+
+        # Create dicts for lmp data and the weight of each cluster
+        rep_days_data = {}
+        weights_data = {}
+
+        rep_days_data = pd.DataFrame(
+            centroids.transpose(), columns=range(1, n_clusters + 1)
+        )
+        lmp_data = rep_days_data.to_dict()
+        weights_data = pd.DataFrame(weights_counter)
+        weights_data.index = np.arange(1, len(weights_data) + 1)
+        weights = weights_data.to_dict()
+
+        return lmp_data, weights
+
+    def append_lmp_data(
+        self,
+        file_path,
+        sheet=None,
+        column_name=None,
+        n_clusters=None,
+        horizon_length=None,
+    ):
+        """
+        This function appends LMP data to the PriceTakerModel using single
+        or multiple year signals, as well as using full price data or
+        clustering to use representative days.
+
+        Args:
+            file_path:      path to the file containing LMP data
+            sheet:          if the file is an excel file, the sheet name
+                            where the LMP data is located - (default: None)
+            column_name:    string name of the column to be used for the
+                            LMP signal
+            n_clusters:     number of clusters for the data (representative days)
+                            None if representative days not used - (default: None)
+            horizon_length: if a value is given, this will be used to set the
+                            horizon_length attribute of the PriceTakerModel.
+                            (default: None --> use existing horizon_length (default: 24))
+
+        Returns:
+
+        """
+        if os.path.exists(file_path):
+            path_to_file = file_path
+        else:
+            raise ValueError(
+                f"The file path {file_path} does not exist. Please check your file path."
+            )
+
+        if ".xls" in path_to_file.suffix:
+            if sheet is None:
+                _logger.warning(
+                    f"Excel file was provided but no sheet was specified. Using the first sheet of the excel file."
+                )
+                sheet = 0
+            full_data = pd.read_excel(path_to_file, sheet_name=[sheet])[sheet]
+        elif ".csv" in path_to_file.suffix:
+            full_data = pd.read_csv(
+                path_to_file,
+            )
+
+        if column_name is None:
```

Comment:
A minor comment, but I would generally put the simple checks first before trying to open the file. These checks are a lot cheaper than opening a file, and I prefer to fail sooner rather than later.